set/name_trie.go

```diff
@@ -0,0 +1,96 @@
+package set
+
+import (
+    "sort"
+)
+
+type NameTrie struct {
+    children   map[rune]*NameTrie
+    terminates bool
+}
+
+func createNameTrie() *NameTrie {
+    return &NameTrie{make(map[rune]*NameTrie), false}
+}
+
+func (tree *NameTrie) Exists(name string) bool {
+    nameSlice := []rune(name)
+    node, ok := tree.traverse(nameSlice, false)
+    return ok && node.terminates
+}
+
+func (tree *NameTrie) Insert(name string) {
+    nameSlice := []rune(name)
+    node, ok := tree.traverse(nameSlice, true)
+    if ok {
+        node.terminates = true
+    }
+}
+
+func (tree *NameTrie) Delete(name string) {
+    nameSlice := []rune(name)
+    node, ok := tree.traverse(nameSlice, false)
+    if ok {
+        node.terminates = false
+    }
+}
+
+func (node *NameTrie) OrderedChildren() []rune {
+    keys := []string{}
+    for suffix := range(node.children){ keys = append(keys, string(suffix)) }
+    sort.Strings(keys)
+    sorted_runes := []rune{}
+    for i := range(keys){ sorted_runes = append(sorted_runes, rune(keys[i][0]))}
+    return sorted_runes
+}
+
+func (tree *NameTrie) traverse(remainder []rune, create bool) (nextTree *NameTrie, ok bool) {
+    nextRune := remainder[0]
+    nextTree, ok = tree.children[nextRune]
+    if !ok {
+        if create {
+            tree.children[nextRune] = createNameTrie()
+            nextTree = tree.children[nextRune]
+        } else {
+            return createNameTrie(), false
+        }
+    }
+    if len(remainder) < 2 {
+        return nextTree, true
+    } else {
+        return nextTree.traverse(remainder[1:], create)
+    }
+}
+
+func (tree *NameTrie) FirstName(name string) (string, bool){
+    nameSlice := []rune(name)
+    node, ok := tree.traverse(nameSlice, false)
+    searchingName, nameFound := "", false
+    if ok {
+        c := make(chan string)
+        go func(){
```

Comment:
Well it won't look like a generator but you'd start with an empty slice and append to it with each result. Nothing fancy. Won't be any slower than making a generator, though. 

Sometimes people implement iterator-like things in Go (typically an interface that has a `.Next()`): https://ewencp.org/blog/golang-iterators/

It's not particularly idiomatic. I wouldn't start with that unless you have something that is highly stateful and incremental and you need to control each incremental step. That's not the case here though, we're just returning the full result each time.